src/daemon/distribution.c

```diff
@@ -0,0 +1,280 @@
+/**
+ * collectd - src/daemon/distribution.c
+ * Copyright (C) 2019-2020  Google LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Svetlana Shmidt <sshmidt at google.com>
+ **/
+
+#include "distribution.h"
+
+#include <errno.h>
+#include <math.h>
+#include <stdio.h>
+#include <string.h>
+
+struct distribution_s {
+  bucket_t *tree;
+  size_t num_buckets;
+  double total_sum;
+};
+
+/**
+ * This code uses an Euler path to avoid gaps in the tree-to-array mapping.
+ * This way the tree contained N buckets contains 2 * N - 1 nodes
+ * Thus, left subtree has 2 * (mid - left + 1) - 1 nodes,
+ * therefore the right subtree starts at node_index + 2 * (mid - left + 1).
+ * For a detailed explanation, see https://docs.google.com/document/d/1ccsg5ffUfqt9-mBDGTymRn8X-9Wk1CuGYeMlRxmxiok/edit?usp=sharing".
+ */
+
+static size_t left_child_index(size_t node_index,
+                               __attribute__((unused)) size_t left,
+                               __attribute__((unused)) size_t right) {
+  return node_index + 1;
+}
```

Comment:
I would recommend to use mutex. That applies to the all functions that access and modifies data. The collectd daemon is multi-threaded. We don't want to our data change in the time we try to read or write sth.

Reply:
Thanks for your suggestion! I've done it ✅